### PR TITLE
Improved depends_on

### PR DESCRIPTION
**Pull request by pierlauro**
Hi,
I eliminated the arbitrary limit of 100 re-try at containers' start time. It was used to detect a probable circular dependency in the use of `depends_on`.

In this PR:

- Each time the configuration file is parsed, instances are now sorted in a DAG fashion so that - at instance start time - it is possible to determine if there is a circular dependency.
- `singularity-compose down` is shutting down the services in reverse order based on `depends_on`: if A is needed by B (B depends on A), it means that B will be shutted down before A.

### CODE REVIEW

**Comment by pierlauro** [comment]
I have a doubt on `get_instance_names`: to me it seems that `self.instances` is always instantiated, maybe it is worth removing the configuration `elif` branch? 

If the PR looks good I will also update the documentation accordingly.

File: scompose/project/project.py [hunk 1]
```diff
@@ -50,9 +50,11 @@ def get_instance_names(self):
            are defined.
         """
         names = []
-        if self.config is not None:
-            if "instances" in self.config:
-                names = list(self.config["instances"].keys())
+        if self.instances is not None:
+            names = self.instances.keys()
```
Comment (vsoch) [inline scompose/project/project.py]: Would this need to be made a list as well?

File: scompose/project/project.py [hunk 2]
```diff
@@ -50,9 +50,11 @@ def get_instance_names(self):
...
+        elif self.config is not None and "instances" in self.config:
+            names = list(self.config["instances"].keys())
+
```
Comment (vsoch) [inline scompose/project/project.py]: My thinking here is that we should not again be parsing the self.config - the parse() function called on init should already parse instances into self.instances. I don't see a case where we would want to derive names from self.config["instances"] because they should already be in self.instances. Thoughts?

File: scompose/project/project.py [hunk 3]
```diff
@@ -166,6 +168,23 @@ def parse(self):
                     working_dir=self.working_dir,
                 )
 
+            # Eventually reorder instances based on depends_on constraints
+            sorted_instances = []
+            for instance in self.instances:
+                depends_on = self.instances[instance].params.get("depends_on", [])
+
+                try:
+                    index = sorted_instances.index(instance)
```
Comment (vsoch) [inline scompose/project/project.py]: So this is to say that we parse depends on, and if we don't find the instance that it depends on, we add to the sorted_instances list anyway and the index is for that instance instead. Then finally we loop over once more, and put any missing ones at the end. This would mean that the ones without dependencies are removed last - that makes sense.

I think since this is adding a large chunk of code to the initial parse function, and since it could be useful to have this other times to get a list of sorted instances based on dependencies, this would do well to be a separate function called here. It should take in some current set of instances, and return the sorted set. It shouldn't be setting or changing anything for self.instances - this should be done for the calling function. That way, we can also write a test for it.

File: scompose/project/project.py [hunk 4]
```diff
@@ -347,7 +366,11 @@ def down(self, names=None):
            names: a list of names of instances to bring down. If not specified, we
            bring down all instances.
         """
-        names = names or self.get_instance_names()
+        if not names:
+          names = list(self.get_instance_names())
```
Comment (vsoch) [inline scompose/project/project.py]: This might be my eyes, but I don't think this is enough spaces here? (should be 4).

Comment (vsoch) [reply scompose/project/project.py]: Also, get_instance_names should return a list so we don't need to do that here. I think I commented on this above when there was a reference to keys() without list wrapping it.

File: scompose/project/project.py [hunk 5]
```diff
@@ -411,50 +434,38 @@ def _create(
...
-                    # Generate a resolv.conf to bind to the container
-                    if self.sudo and not no_resolv:
-                        resolv = self.generate_resolv_conf()
-                        instance.volumes.append("%s:/etc/resolv.conf" % resolv)
-
-                        # Create a hosts file for the instance based, add as volume
-                        instance.volumes.append("%s:/etc/hosts" % hosts_file)
+        for instance in self.iter_instances(names):
+            depends_on = instance.params.get("depends_on", [])
+            for dep in depends_on:
```
Comment (vsoch) [inline scompose/project/project.py]: I don't understand this logic here - if a dependency isn't created yet, how can you be sure that it's circular?

File: scompose/project/project.py [hunk 6]
```diff
@@ -411,50 +434,38 @@ def _create(
...
+            # Generate a resolv.conf to bind to the container
+            if self.sudo and not no_resolv:
+                resolv = self.generate_resolv_conf()
+                instance.volumes.append("%s:/etc/resolv.conf" % resolv)
+
+                # Create a hosts file for the instance based, add as volume
+                instance.volumes.append("%s:/etc/hosts" % hosts_file)
+
+            # If we get here, execute command and add to list
+            create_func = getattr(instance, command)
```
Comment (vsoch) [inline scompose/project/project.py]: Could you walk me through the logic of this code block?

File: scompose/project/project.py [hunk 7]
```diff
@@ -411,50 +434,38 @@ def _create(
...
-                    names.remove(instance.name)
+            # Run post create commands
+            instance.run_post()
 
-                    # Run post create commands
-                    instance.run_post()
+        if circular_dep:
+            bot.exit(
+                "Unable to create all instances, possible circular dependency."
+            )
```
Comment (vsoch) [inline scompose/project/project.py]: I definitely think it's improved to not use a count :)

File: scompose/version.py [hunk 8]
```diff
@@ -8,7 +8,7 @@
 
 """
 
-__version__ = "0.0.17"
+__version__ = "0.0.18"
 AUTHOR = "Vanessa Sochat"
```
Comment (vsoch) [inline scompose/version.py]: Great job doing this! We would want to add a note to the CHANGELOG.md in the root of the repository as well.

**Comment by vsoch** [comment]
> I have a doubt on get_instance_names: to me it seems that self.instances is always instantiated, maybe it is worth removing the configuration elif branch?

Oh yes, this is spot on! I realized this during my review and it looks like you picked up on it too. +1.

File: scompose/project/project.py (hunk 3 above)
Comment (pierlauro) [reply scompose/project/project.py]: > this should be done for the calling function. That way, we can also write a test for it.

Makes total sense, I am gonna do it!

File: scompose/project/project.py (hunk 5 above)
Comment (pierlauro) [reply scompose/project/project.py]: Because the `instances` list is sorted out in such a way that if the dependencies are not a DAG, at least one dependency will be missing.

Let's take this singularity-compose file as an example:
```
version: "1.0"
instances:
  client0:
    build:
      context: ./images
    depends_on: [client2]
  client1:
    build:
      context: ./images
    depends_on: [client2]
  client2:
    build:
      context: ./images
```
The instances list will be sorted out in this way `[client2, client0, client1]`, so that `client2` will be created before the dependent containers.


Let's consider a wrong singularity-compose file instead:
```
version: "1.0"
instances:
  client0:
    build:
      context: ./images
    depends_on: [client2]
  client1:
    build:
      context: ./images
    depends_on: [client2]
  client2:
    build:
      context: ./images
      depends_on: [client0]
```
The instances list will be sorted out in this way `[client0, client1, client2]`, so that `client0` will not run is `client2` is not up already: that will raise the circular dependency error.

Comment (pierlauro) [reply scompose/project/project.py]: The sorting happens [on this line](https://github.com/singularityhub/singularity-compose/blob/894e4eb2ff2faed44ba027bacde676d0c065cb6b/scompose/project/project.py#L171)

File: scompose/project/project.py (hunk 6 above)
Comment (pierlauro) [reply scompose/project/project.py]: It was already like that, I just reformatted the spacing once deleted the outer loop.

I looked into it, the passed `command` is the name of the method to execute on the `instance`: can be either `create` (in case of build) or simply `up` (in case of run)

File: scompose/project/project.py (hunk 3 above)
Comment (pierlauro) [reply scompose/project/project.py]: Would you see it better as Project's method (as in f864a1d) or as a standalone method in utils?

File: scompose/project/project.py [hunk 9]
```diff
@@ -51,9 +51,7 @@ def get_instance_names(self):
         """
         names = []
         if self.instances is not None:
-            names = self.instances.keys()
-        elif self.config is not None and "instances" in self.config:
-            names = list(self.config["instances"].keys())
+            names = list(self.instances.keys())
```
Comment (vsoch) [inline scompose/project/project.py]: perfect.

File: scompose/project/project.py (hunk 3 above)
Comment (vsoch) [reply scompose/project/project.py]: Let's think about this - so the project instance is meant to load and represent a project file (singularity-compose.yml) and doesn't even load without it. Currently we use utils for functions to read / write / otherwise interact with yaml or files, so my thinking is that I don't see any obvious use cases where we would want to have the function outside of the project file - using the Project class always helps the user by handling the load and some parsing, so my thinking is that it should be a project method. What are your thoughts? Are there use cases for having it external that I'm not thinking of?

File: scompose/project/project.py (hunk 5 above)
Comment (vsoch) [reply scompose/project/project.py]: ah okay - do we have an easy way to add a simple test with a file like this?

File: scompose/project/project.py (hunk 6 above)
Comment (vsoch) [reply scompose/project/project.py]: Ah okay, what I was curious about is how the logic works with the deleted outer loop. For example, the do_create variable is gone, If there is a way to test this and make sure that the functionality has not changed, then that should be enough validation for approval.

**Comment by pierlauro** [comment]
I just tested manually the over-mentioned features (sorting based on `depends_on`, circular dependencies and eliminate outer loop with `do_create`). 

I could easily test the DAG sorting function and I am gonna do that.

Regarding the other parts - instead - right now there is no easy way to mock components (e.g. mock containers creation/execution without using singularity for real) and setting that up seems a lengthy task detached from this pull request. ATM I would rely on the manual checks I did, what are your feelings about that?

**Comment by vsoch** [comment]
Well, if you look at the [circle setup](https://github.com/singularityhub/singularity-compose/blob/master/.circleci/config.yml) and [tests](https://github.com/singularityhub/singularity-compose/blob/master/scompose/tests/test_client.py#L45) that are currently run, the hard part to have singularity is already handled. From this starting point, wouldn't it be easy to add another config file to the tests folder, and bring up a small busybox group that would each (as their startscript) write their own name to a file, then we could read the file and confirm that the order is correct?

**Comment by vsoch** [comment]
If you are exhausted I could take a shot as well - and I'd open with your commits in another PR so the testing could run (as opposed to PR to your branch which wouldn't run). 

I'll be back at computer in maybe 4-5 hours (still 6am here!) and would be happy to help :)

**Comment by pierlauro** [comment]
> Well, if you look at the [circle setup](https://github.com/singularityhub/singularity-compose/blob/master/.circleci/config.yml) and [tests](https://github.com/singularityhub/singularity-compose/blob/master/scompose/tests/test_client.py#L45) that are currently run, the hard part to have singularity is already handled. From this starting point, wouldn't it be easy to add another config file to the tests folder, and bring up a small busybox group that would each (as their startscript) write their own name to a file, then we could read the file and confirm that the order is correct?

Oh, perfect! Let me look into it during the weekend as I am working on something else ATM :)

**Comment by vsoch** [comment]
Please see #32 hopefully I got us most of the way there!

**Comment by pierlauro** [comment]
> Please see #32 hopefully I got us most of the way there!

I've merged your PR in this one since I had to test the changes. I've created a failing test (with a circular dependency) and a successful one (also checking correctness of the "DAG list")

**Comment by vsoch** [comment]
@pierlauro it's good practice to respect work by others - you've taken my work and added via commits under your name. Please either pull those commits, or work from my branch, I won't be merging something that does not correctly give attribution to the correct author.

**Comment by vsoch** [comment]
> I've merged your PR in this one since I had to test the changes.

I do not see my commits represented here.

File: scompose/project/project.py (hunk 3 above)
Comment (pierlauro) [reply scompose/project/project.py]: Yeah, it makes sense to keep as Project's method, I've also indirectly tested it in the new commit.

**Comment by pierlauro** [comment]
My bad, as long as the software is enhanced I usually don't care about attribution but I understand it can be different for other people.
I had copied your files as github's web interface doesn't allow to open a PR from the original repo (your `depends_on` branch) to a forked one (my `depends_on` branch) but just the opposite. As you raised the issue, how would you recommend to proceed? (Just in case, I've anyway saved the last changed in the branch `depends_on_bak` of my repo).

**Comment by pierlauro** [comment]
> My bad, as long as the software is enhanced I usually don't care about attribution but I understand it can be different for other people.
> I had copied your files as github's web interface doesn't allow to open a PR from the original repo (your `depends_on` branch) to a forked one (my `depends_on` branch) but just the opposite. As you raised the issue, how would you recommend to proceed? (Just in case, I've anyway saved the last changed in the branch `depends_on_bak` of my repo).

Ok, I merged through the CLI. Give me a couple of minutes to replicate my changes

File: scompose/tests/test_depends_on.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    print("Testing build")
+    project.build()
+
+    for image in ["first.sif", "second.sif", "third.sif"]:
+        assert image in os.listdir(tmp_path)
+
+    print("Testing view config")
+    project.view_config()
+
+    # Test depends_on DAG order
```
Comment (vsoch) [inline scompose/tests/test_depends_on.py]: The one addition I was thinking would make sense to test is to be able to have the instances echo / write to some shared file about the order they are starting up in (that's why the startscripts have the echos for now). What do you think?

File: scompose/project/project.py [hunk 11]
```diff
@@ -347,7 +367,11 @@ def down(self, names=None):
            names: a list of names of instances to bring down. If not specified, we
            bring down all instances.
         """
-        names = names or self.get_instance_names()
+        if not names:
+            names = self.get_instance_names()
+            # Ordered shutdown in case of depends_on
+            names.reverse()
```
[the remark this answers is not in the document]
Comment (pierlauro) [reply scompose/project/project.py]: That would have been totally right in python 2, I originally used `OrderedDict` in my solution but then discovered that in Python 3 keys are [sorted in insertion order](https://docs.python.org/3/library/stdtypes.html#dict) (that is actually what we're looking for in this case)

File: scompose/project/project.py [hunk 12]
```diff
@@ -166,10 +166,30 @@ def parse(self):
...
+                index = sorted_instances.index(instance)
+            except ValueError:
+                sorted_instances.append(instance)
+                index = sorted_instances.index(instance)
+
+            for dep in depends_on:
+                if not dep in sorted_instances:
+                    sorted_instances.insert(index, dep)
+
+        return {k: self.instances[k] for k in sorted_instances}
```
[the remark this answers is not in the document]
Comment (pierlauro) [reply scompose/project/project.py]: Uhm, that way you are saying that whenever the function is called without names, we should provide the instances in reverse order. Isn't it a risky assumption? I would rather let instances always to be correctly ordered based on `depends_on` and just delegate the reverse to the `down` function.   

Regarding `OrderedDict`, I just replied in the next comment.

File: .docs/changelog.md [hunk 13]
```diff
@@ -1 +0,0 @@
-../CHANGELOG.md
```
[the remark this answers is not in the document]
Comment (pierlauro) [reply .docs/changelog.md]: Isn't it actually still in the branch?

[the remark this answers is not in the document]
Comment (pierlauro) [reply .docs/changelog.md]: Isn't it actually still in the branch?

File: scompose/tests/test_depends_on.py (hunk 10 above)
Comment (pierlauro) [reply scompose/tests/test_depends_on.py]: Oh ok, so your idea is to test the output rather than the instances list in case the internal code would change, right?

Comment (vsoch) [reply scompose/tests/test_depends_on.py]: I think we might do well to test the instances list (as you've done) and the actual output, in case the way we have them listed doesn't correspond to the start order.

File: scompose/project/project.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (vsoch) [reply scompose/project/project.py]: Cool, I didn't realize that! I think not using the OrderedDict is cleaner, I'm okay with not using it (and hoping users have at least 3.7) and we can address the issue if it comes up in the wild.

File: scompose/project/project.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (vsoch) [reply scompose/project/project.py]: I didn't say to add names.reverse() to the function to get the names, I suggested just using OrderedDict, so _after_ that is called, when we call reverse() it would be in reverse order. But you've clarified that for Python 3.7 and above, the dictionary is already honored to be ordered by insertion, so the reverse() would work fine here. Does this clear up your misread of my comment?

File: .docs/changelog.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (vsoch) [reply .docs/changelog.md]: You deleted .docs/changelog.md, we still have `CHANGELOG.md`. The .docs/changelog.md makes sure that https://github.com/singularityhub/singularity-compose/blob/master/CHANGELOG.md renders a change log into the site.

File: scompose/project/project.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (pierlauro) [reply scompose/project/project.py]: > Does this clear up your misread of my comment?

Yeah, definitely! I had wrongly associated your comment to the quoted piece of code

File: scompose/logger/message.py [hunk 14]
```diff
@@ -281,6 +281,10 @@ def table(self, rows, col_width=2):
             message = "\t".join(row)
             self.custom(prefix=label, message=message)
 
+    def clear(self):
```
Comment (vsoch) [inline scompose/logger/message.py]: I like this!

**Comment by vsoch** [comment]
we should be good to go! The CI (CircleCI) wasn't triggered to run on a pull request, so I updated the API docs and enabled building forks, and after that passes we should be good to go.

File: scompose/logger/message.py (hunk 14 above)
Comment (pierlauro) [reply scompose/logger/message.py]: Much cleaner than the [usual pytest way of intercepting stdout](https://docs.pytest.org/en/latest/capture.html), very good idea to have that simple logger!

**Comment by vsoch** [comment]
okay, so the Singularity install is working ok (and I've updated to use more recent versions) what it looks like is that the runner on CircleCI is expecting a Singularity recipe file (and not honoring Singularity.first, etc.). I'm not sure why, especially because the test_client.py works (which does a similar build of https://github.com/singularityhub/singularity-compose-examples/tree/master/apache-simple, and there isn't a Singularity file there.

**Comment by pierlauro** [comment]
> okay, so the Singularity install is working ok (and I've updated to use more recent versions) what it looks like is that the runner on CircleCI is expecting a Singularity recipe file (and not honoring Singularity.first, etc.). I'm not sure why, especially because the test_client.py works (which does a similar build of https://github.com/singularityhub/singularity-compose-examples/tree/master/apache-simple, and there isn't a Singularity file there.

The build step requires sudo privileges, could it be a problem of permissions?

**Comment by vsoch** [comment]
I think it could be - that's the only difference that I can produce locally. Before we dig into that, I just want to do some print outs in the CI for a sanity check (it does work for me locally). Here is a local run:

```
Copying Singularity.second to /tmp/pytest-of-vanessa/pytest-16/test_circular_dependency0/Singularity.second
Copying singularity-compose.yml to /tmp/pytest-of-vanessa/pytest-16/test_circular_dependency0/singularity-compose.yml
Copying Singularity.first to /tmp/pytest-of-vanessa/pytest-16/test_circular_dependency0/Singularity.first
Creating project...
DEBUGGING PARAMS
Singularity.first
{'build': {'context': '.', 'recipe': 'Singularity.first'}, 'depends_on': ['second']}
PRESENT WORKING DIRECTORY
/tmp/pytest-of-vanessa/pytest-16/test_circular_dependency0
['Singularity.third', 'Singularity.second', 'singularity-compose.yml', 'Singularity.first']
DEBUGGING PARAMS
Singularity.second
{'build': {'context': '.', 'recipe': 'Singularity.second'}, 'depends_on': ['second']}
PRESENT WORKING DIRECTORY
/tmp/pytest-of-vanessa/pytest-16/test_circular_dependency0
['Singularity.third', 'Singularity.second', 'singularity-compose.yml', 'Singularity.first']
DEBUGGING PARAMS
Singularity.third
{'build': {'context': '.', 'recipe': 'Singularity.third'}}
PRESENT WORKING DIRECTORY
/tmp/pytest-of-vanessa/pytest-16/test_circular_dependency0
['Singularity.third', 'Singularity.second', 'singularity-compose.yml', 'Singularity.first']
```
and I'll post the CI run here when I get it. What we *could* try (not sure if it would be supported) is --fakeroot on the CI. What isn't clear to me is why the self.recipe (which should be Singularity.first, second, third, etc. is just being parsed as "Singularity"). Another possibility is if instances / tests are being run in parallel, it could be that messes something up.

Oh wait, I think I just reproduced locally - it's because of the command to pytest (and perhaps where the $PWD is?) Let me try again - I think it might be because pytest is running things at the same time, and it's messing it up.

**Comment by pierlauro** [comment]
>  I think it might be because pytest is running things at the same time, and it's messing it up.

Good guess! If tests are run in parallel, both the `depends_on` ones are sharing the same container name and messing things up

**Comment by vsoch** [comment]
Woot, that did the trick! The giveaway was seeing unexpected output for another test mixed with the tests we wrote. And then on top of this issue, the circle config was (for some reason) no longer valid, so I was working on it locally:

```bash
circleci config validate .circleci/config.yml 
```
And updated the usage for the singularity orb, and also enforced running the tests one at a time by way of calling pytest more than once. And whew, all is working as expected! I also simplified the config quite a bit - I always feel clever at points when I do some wacky, advanced yaml nonsense, and then years later, usually regret it (and like I did now, realize that the simplest structure is going to work the best!)

@pierlauro such great work you did here, and now we are tested and green! It's time for merge! I'll post the package release here shortly after. Thank you so kindly for working on this!

**Comment by vsoch** [comment]
https://pypi.org/project/singularity-compose/0.0.18/

**Comment by pierlauro** [comment]
> Woot, that did the trick! The giveaway was seeing unexpected output for another test mixed with the tests we wrote. And then on top of this issue, the circle config was (for some reason) no longer valid, so I was working on it locally:
> 
> ```shell
> circleci config validate .circleci/config.yml 
> ```
> 
> And updated the usage for the singularity orb, and also enforced running the tests one at a time by way of calling pytest more than once. And whew, all is working as expected! I also simplified the config quite a bit - I always feel clever at points when I do some wacky, advanced yaml nonsense, and then years later, usually regret it (and like I did now, realize that the simplest structure is going to work the best!)
> 
> @pierlauro such great work you did here, and now we are tested and green! It's time for merge! I'll post the package release here shortly after. Thank you so kindly for working on this!

Great! Thank you very much for your time on that!
